onnxruntime/contrib_ops/cpu/bert/attention.cc

```diff
@@ -219,44 +357,45 @@ Status Attention<T>::PrePack(const Tensor& weights, int input_idx, AllocatorPtr
+  size_t q_hidden_size, k_hidden_size, v_hidden_size;
 
-  packed_weights_size_ = MlasGemmPackBSize(head_size, input_hidden_size);
-  if (packed_weights_size_ == 0) {
-    return Status::OK();
-  }
+  if (qkv_hidden_sizes_.size() != 0) {
+    q_hidden_size = qkv_hidden_sizes_[0];
+    k_hidden_size = qkv_hidden_sizes_[1];
+    v_hidden_size = qkv_hidden_sizes_[2];
 
-  const size_t loop_len = static_cast<size_t>(3) * num_heads_;
-  size_t packed_weights_data_size = packed_weights_size_ * loop_len;  // The same size would be computed by AllocArray() below
-  auto* packed_weights_data = static_cast<uint8_t*>(alloc->AllocArray(packed_weights_size_, loop_len));
+    if (q_hidden_size == 0 || k_hidden_size == 0 || v_hidden_size == 0) {
+      return Status::OK();
+    }
 
-  // Initialize memory to 0 as there could be some padding associated with pre-packed
-  // buffer memory and we don not want it uninitialized and generate different hashes
-  // if and when we try to cache this pre-packed buffer for sharing between sessions.
-  memset(packed_weights_data, 0, packed_weights_data_size);
+    if (q_hidden_size % num_heads_ != 0 || k_hidden_size % num_heads_ != 0 || v_hidden_size % num_heads_ != 0) {
+      return Status::OK();
+    }
+  } else {
+    const size_t hidden_size_x3 = static_cast<size_t>(weights_dims[1]);
+    const size_t hidden_size = hidden_size_x3 / 3;
 
-  packed_weights_ = BufferUniquePtr(packed_weights_data, BufferDeleter(alloc));
+    if (hidden_size % num_heads_ != 0) {
+      return Status::OK();
+    }
 
-  for (size_t i = 0; i < loop_len; i++) {
-    MlasGemmPackB(CblasNoTrans, head_size, input_hidden_size, weights_data, hidden_size_x3, packed_weights_data);
-    packed_weights_data += packed_weights_size_;
-    weights_data += head_size;
+    q_hidden_size = hidden_size;
+    k_hidden_size = hidden_size;
+    v_hidden_size = hidden_size;
   }
 
-  bool share_prepacked_weights = (prepacked_weights != nullptr);
-  if (share_prepacked_weights) {
-    prepacked_weights->buffers_.push_back(std::move(packed_weights_));
-    prepacked_weights->buffer_sizes_.push_back(packed_weights_data_size);
+  const size_t q_head_size = q_hidden_size / num_heads_;
+  const size_t k_head_size = k_hidden_size / num_heads_;
+  const size_t v_head_size = v_hidden_size / num_heads_;
+  const size_t weight_matrix_col_size = q_hidden_size + k_hidden_size + v_hidden_size;
+
+  if (!IsPackWeightsSuccessful(0, alloc, q_head_size, input_hidden_size, weights_data, weight_matrix_col_size, prepacked_weights) ||
+      !IsPackWeightsSuccessful(1, alloc, k_head_size, input_hidden_size, weights_data + (num_heads_ * q_head_size), weight_matrix_col_size, prepacked_weights) ||
+      !IsPackWeightsSuccessful(2, alloc, v_head_size, input_hidden_size, weights_data + (num_heads_ * (q_head_size + k_head_size)), weight_matrix_col_size, prepacked_weights)) {
+    // we are not cleaning up anything, assuming caller takes care of this
+    return Status::OK();
```

Comment:
Do kernel class members set in any successful call of `IsPackWeightsSuccessful()` need to be cleaned up/reset ? For example, is it possible that `q_packed_weights_ ` be non-null (successful) but `k_packed_weights_`\`v_packed_weights_` be null (unsuccessful) ? If so, is that a valid state for the instance of the kernel ? On line 537, the check is on `q_packed_weights_` and if it is non-null, `k...` and `q`... are inherently assumed to be non-null as well. Is that safe ?

The comment seems to hint at clean-up of PrePackedWeights instance. This is true as long `is_packed` is not set to true as the PrePackedWeights instance will go out of scope in `SessionState` thus de-allocating any buffers it held at that point...

Reply:
very good point, discussed offline.

1. since the buffers are split, we need to handle the case of all k, v failing after q or q,k are allocated. I will add another variable which helps us with this. 
2.  Shared case and non-shared case memory would be differently de-allocated in case of failure. In shared case, the pointers are moved from kernel and never initiated again. In non-shared case, the pointer is held with kernel and never used until the point of completion.